### PR TITLE
core/consensus: add ShouldRun method

### PR DESCRIPTION
**Pull request by dB2510**
Adds a `ShouldRun` method to `instanceIO` to check if the instance was actually "running" rather than returning boolean from `getInstanceIO`.

category: bug
ticket: #2439

### CODE REVIEW

**Review by corverroos — APPROVED** [review]

**Comment by corverroos** [comment]
A test that calls the handle function with some data before Propose/Participate would be great

File: core/consensus/component.go [hunk 1]
```diff
@@ -177,6 +179,19 @@ func (io *instanceIO) MarkProposed() error {
 	return nil
 }
 
+// ShouldRun marks the instance as running.
```
Comment (xenowits) [inline core/consensus/component.go:182]:
```suggestion
// ShouldRun returns true and marks the instance as running if the instance was not already running.
```

File: core/consensus/component.go [hunk 2]
```diff
@@ -177,6 +179,19 @@ func (io *instanceIO) MarkProposed() error {
...
+// It returns false if the instance was already marked as running.
```
Comment (xenowits) [inline core/consensus/component.go:183]:
```suggestion
// ShouldRun checks the current status of the instance.
// If the instance is not already running, it returns true and marks the instance as running.
// If the instance is already running, it returns false.
```

**Review by xenowits — APPROVED** [review]

File: core/consensus/component_internal_test.go [hunk 3]
```diff
@@ -454,6 +454,73 @@ func TestComponentHandle(t *testing.T) {
...
+
+	t.Run("ShouldRun after handle", func(t *testing.T) {
+		var c Component
+		c.deadliner = testDeadliner{}
+		c.mutable.instances = make(map[core.Duty]instanceIO)
+
+		msg := &pbv1.ConsensusMsg{
+			Msg: randomMsg(t),
+		}
+		p2pKey := testutil.GenerateInsecureK1Key(t, 0)
```
Comment (xenowits) [inline core/consensus/component_internal_test.go:472]: nit: add newline

File: core/consensus/component_internal_test.go [hunk 4]
```diff
@@ -454,6 +454,73 @@ func TestComponentHandle(t *testing.T) {
...
+		c.pubkeys = make(map[int64]*k1.PublicKey)
+		c.pubkeys[0] = p2pKey.PubKey()
+
+		duty := core.Duty{Slot: 42, Type: 1}
+		msg = signConsensusMsg(t, msg, p2pKey, duty)
+
+		// It should create new instance of instanceIO for the give duty.
```
Comment (xenowits) [inline core/consensus/component_internal_test.go:479]:
```suggestion
		// It should create new instance of instanceIO for the given duty.
```

File: core/consensus/component_internal_test.go [hunk 5]
```diff
@@ -454,6 +454,73 @@ func TestComponentHandle(t *testing.T) {
...
+		_, _, err := c.handle(ctx, "peerID", msg)
+		require.NoError(t, err)
+
+		pubkey := testutil.RandomCorePubKey(t)
+
+		// It should mark instance as running by calling inst.ShouldRun().
```
Comment (xenowits) [inline core/consensus/component_internal_test.go:513]:
```suggestion
		// Propose should internally mark instance as running by calling inst.ShouldRun().
```

**Review by xenowits — APPROVED** [review]

**Pull request merged**